### PR TITLE
Support of Istio Client API

### PR DESCRIPTION
**Pull request by Sgitario**
This PR supports v1alpha3 and v1beta1 Istio API
Solution is based on https://github.com/snowdrop/istio-java-api
Tests have been moved from the above repository to here.
Fixes https://github.com/fabric8io/kubernetes-client/issues/3593

Apart from adding the Istio API, we added two changes to the GO generator:
- Ability to extract the field name from the protobug tag: this is optional and only enabled by a flag
- Fix mapping of float numbers
These changes are in separated commits.

### CODE REVIEW

**Comment by Sgitario** [comment]
fyi @metacosm

**Comment by rohanKanojia** [comment]
Have you copy pasted code from Istio Java API? Please preserve commit history https://github.com/fabric8io/kubernetes-client/pull/1605#pullrequestreview-256033346

**Comment by Sgitario** [comment]
> Have you copy-pasted code from Istio Java API? Please preserve commit history We did something similar while porting Service Catalog extension [#1605 (review)](https://github.com/fabric8io/kubernetes-client/pull/1605#pullrequestreview-256033346)

Didn't copy paated code from the snowdrop repo? I used this repo as a reference and also migrated the tests to here.

File: extensions/istio/generator-v1alpha3/go.mod [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module github.com/fabric8io/kubernetes-client/extensions/istio/generator
+
+go 1.13
+
+require (
+	github.com/fabric8io/kubernetes-client/generator v0.0.0
+	github.com/gogo/protobuf v1.3.2
+	istio.io/api v0.0.0-20211111170902-f2972da3100e
+	istio.io/client-go v1.12.0-alpha.5.0.20211111180919-207fea46070b
```
Comment (metacosm) [inline extensions/istio/generator-v1alpha3/go.mod:9]: Can we use the released version, instead, please?

Comment (Sgitario) [reply extensions/istio/generator-v1alpha3/go.mod]: Updated

File: extensions/istio/model-v1alpha3/src/main/java/io/fabric8/istio/api/networking/v1alpha3/CorsPolicy.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  @BuildableReference(ObjectMeta.class),
+  @BuildableReference(ObjectReference.class),
+  @BuildableReference(LabelSelector.class),
+  @BuildableReference(Container.class),
+  @BuildableReference(EnvVar.class),
+  @BuildableReference(ContainerPort.class),
+  @BuildableReference(Volume.class),
+  @BuildableReference(VolumeMount.class)
+})
+public class CorsPolicy implements KubernetesResource {
```
Comment (manusa) [inline extensions/istio/model-v1alpha3/src/main/java/io/fabric8/istio/api/networking/v1alpha3/CorsPolicy.java:68]: Out of curiosity, what's the deal with this class (in every model version).
Why is it manually created?
Is it not generated automatically? (If it is, why isn't there a delete rule in the pom.xml?)

Comment (Sgitario) [reply extensions/istio/model-v1alpha3/src/main/java/io/fabric8/istio/api/networking/v1alpha3/CorsPolicy.java:68]: In summary, it's because of this sundrio issue: https://github.com/sundrio/sundrio/issues/170
I commented it out in the generator program: https://github.com/fabric8io/kubernetes-client/pull/3613/files#diff-bc93f30af530888e6449c6e0f76d911c332098413e16b51074b62d191560e7edR58

I didn't want to include the comment in the java class because users would see it.

Comment (manusa) [reply extensions/istio/model-v1alpha3/src/main/java/io/fabric8/istio/api/networking/v1alpha3/CorsPolicy.java:68]: Cool.
I think in the main model we don't have this `providedTypes` field, that's why it's confusing. The approach there is to delete the generated type and persist a manually tweaked one.
We should definitely merge this approach there.

**Review by manusa — APPROVED** [review]
LGTM, with some comments:

- I think we can fix the most of the test related code smells
- Some tests are failing on Windows

**Comment by Sgitario** [comment]
FYI, I'm investigating the issue with windows.

**Comment by Sgitario** [comment]
Fixed the Windows check

**Comment by rbaeurle** [comment]
I could verify that this PR seems to work fine for us.

**Review by metacosm — APPROVED** [review]
Great job overall!
A few comments, though:

-  We probably need to document how to migrate to this new version since some method names have changed and, obviously, the packages are now different
- This PR only covers the networking part of the Istio API. While this should cover quite a bit of what people use, there are several other APIs that would need to be ported/updated to match what was available in the `istio-java-api` project (mesh, policy, rbac, security and mixer APIs are missing). Not all of them might need to be ported since some of these are actually obsolete but this is still something that needs to be investigated.
- The `istio-java-api` project also provided a CEL (https://opensource.google/projects/cel) parser via ANTLR to support validation of some of the configuration fields. Relevance to this new version needs to be assessed as well.
- Basically, we need someone who routinely uses Istio to support this API to make sure that the relevant parts are covered correctly.

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import okhttp3.OkHttpClient;
+
+public class V1beta1APIGroupExtensionAdapter extends APIGroupExtensionAdapter<V1beta1APIGroupDSL> {
+  @Override
+  protected String getAPIGroupName() {
+    return "v1beta1";
+  }
+
+  @Override
+  public Class<V1beta1APIGroupDSL> getExtensionType() {
```
Comment (rohanKanojia) [inline extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java:29]:
```suggestion
  public Class<V1alpha3APIGroupClient> getExtensionType() {
```

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    return V1beta1APIGroupDSL.class;
```
Comment (rohanKanojia) [inline extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java:30]:
```suggestion
    return V1alpha3APIGroupClient.class;
```

File: extensions/istio/tests/src/test/java/io/fabric8/istio/test/v1alpha3/DestinationRuleTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.istio.test.v1alpha3;
+
+import static org.junit.Assert.assertTrue;
```
Comment (rohanKanojia) [inline extensions/istio/tests/src/test/java/io/fabric8/istio/test/v1alpha3/DestinationRuleTest.java:18]: We're using Junit5, let's keep all assertions consistent and use Junit5 assertions everywhere
```suggestion
import static org.junit.jupiter.api.Assertions.assertTrue;
```

**Review by rohanKanojia — APPROVED** [review]

**Comment by rohanKanojia** [comment]
There doesn't seem to be any CRUD test for any resource. Would be nice to add some Crud tests to test suite as well.

**Comment by rohanKanojia** [comment]
Sorry for not being clear, I was referring to CRUD mode in MockServer. You don't need to add expectations in this mode. Here is an example:
https://github.com/fabric8io/kubernetes-client/blob/master/extensions/camel-k/tests/src/test/java/io/fabric8/camelk/test/crud/IntegrationCrudTest.java#L29

**Comment by manusa** [comment]
Sorry, I didn't read the comments, before merging (just saw the approval).

Could you add the requested changes to a new PR?

File: README.md [hunk 6]
```diff
@@ -561,6 +561,7 @@ All the resource objects used here will be according to OpenShift 3.9.0  and Kub
   - [Knative](https://github.com/fabric8io/kubernetes-client/tree/master/extensions/knative)
   - [Tekton](https://github.com/fabric8io/kubernetes-client/tree/master/extensions/tekton)
   - [Volcano](https://github.com/fabric8io/kubernetes-client/tree/master/extensions/volcano)
+  - [Istio](https://github.com/fabric8io/kubernetes-client/tree/master/extensions/istio)
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply README.md:564]: Updated

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/DefaultIstioClient.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  public FunctionCallable<NamespacedIstioClient> withRequestConfig(RequestConfig requestConfig) {
+    return new WithRequestCallable<>(this, requestConfig);
+  }
+
+  @Override
+  public V1alpha3APIGroupDSL v1alpha3() {
+    return adapt(V1alpha3APIGroupDSL.class);
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/DefaultIstioClient.java:61]: Fixed

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/DefaultIstioClient.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  public V1beta1APIGroupDSL v1beta1() {
+    return adapt(V1beta1APIGroupDSL.class);
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/DefaultIstioClient.java:66]: Fixed

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1alpha3APIGroupDSL.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import io.fabric8.kubernetes.client.dsl.MixedOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+
+/**
+ * Main interface for Istio Client.
+ */
+public interface V1alpha3APIGroupDSL extends Client {
+
+  // networking
+  MixedOperation<DestinationRule, DestinationRuleList, Resource<DestinationRule>> destinationRules();
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/V1alpha3APIGroupDSL.java:42]: added

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1alpha3APIGroupExtensionAdapter.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.istio.client;
+
+import io.fabric8.kubernetes.client.APIGroupExtensionAdapter;
+import io.fabric8.kubernetes.client.Client;
+import okhttp3.OkHttpClient;
+
+public class V1alpha3APIGroupExtensionAdapter extends APIGroupExtensionAdapter<V1alpha3APIGroupDSL> {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/V1alpha3APIGroupExtensionAdapter.java:22]: changed

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.istio.client;
+
+import io.fabric8.kubernetes.client.APIGroupExtensionAdapter;
+import io.fabric8.kubernetes.client.Client;
+import okhttp3.OkHttpClient;
+
+public class V1beta1APIGroupExtensionAdapter extends APIGroupExtensionAdapter<V1beta1APIGroupDSL> {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java:22]: changed

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java (hunk 3 above)
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java:29]: changed

File: extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java (hunk 4 above)
Comment (Sgitario) [reply extensions/istio/client/src/main/java/io/fabric8/istio/client/V1beta1APIGroupExtensionAdapter.java:30]: changed

File: extensions/istio/tests/src/test/java/io/fabric8/istio/test/v1alpha3/DestinationRuleTest.java (hunk 5 above)
Comment (Sgitario) [reply extensions/istio/tests/src/test/java/io/fabric8/istio/test/v1alpha3/DestinationRuleTest.java:18]: ops, thanks for spotting!

**Comment by Sgitario** [comment]
> Sorry for not being clear, I was referring to CRUD mode in MockServer. You don't need to add expectations in this mode. Here is an example: https://github.com/fabric8io/kubernetes-client/blob/master/extensions/camel-k/tests/src/test/java/io/fabric8/camelk/test/crud/IntegrationCrudTest.java#L29

ok ok! I've added the crud flag to true in the v1beta1 VirtualService test.

File: extensions/istio/client/pom.xml [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-migrationsupport</artifactId>
+      <scope>test</scope>
+    </dependency>
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/istio/client/pom.xml:75]: Removed

**Comment by Sgitario** [comment]
Request changes fixed in https://github.com/fabric8io/kubernetes-client/pull/3650
